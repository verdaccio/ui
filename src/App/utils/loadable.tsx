import React, { lazy, Suspense } from 'react';
export default (importCallback: any) => {
  const TargetComponent = lazy(importCallback);
  return (props: any) => (
    <Suspense fallback={null}>
      <TargetComponent {...props} />
    </Suspense>
  );
};
