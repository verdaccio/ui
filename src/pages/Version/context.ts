import { createContext, Consumer, Provider } from 'react';

import { DetailContextProps, VersionPageConsumerProps } from './version-config';

export const DetailContext = createContext<Partial<DetailContextProps>>({});

export const DetailContextProvider: Provider<Partial<VersionPageConsumerProps>> = DetailContext.Provider;
export const DetailContextConsumer: Consumer<Partial<VersionPageConsumerProps>> = DetailContext.Consumer;
