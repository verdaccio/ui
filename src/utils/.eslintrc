{
    "rules": {
        "no-invalid-this": 0
    }
}