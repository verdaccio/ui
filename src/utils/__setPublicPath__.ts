// @ts-ignore
if (!__DEBUG__) {
  // @ts-ignore
  __webpack_public_path__ = window.VERDACCIO_API_URL.replace(/\/verdaccio\/$/, '/static/'); // eslint-disable-line
}
