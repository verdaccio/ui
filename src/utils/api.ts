import storage from './storage';
import '../../types';

/**
 * Handles response according to content type
 * @param {object} response
 * @returns {promise}
 */
export function handleResponseType(response: Response): Promise<[boolean, any]> {
  if (response.headers) {
    const contentType = response.headers.get('Content-Type');
    if (contentType && contentType.includes('application/pdf')) {
      return Promise.all([response.ok, response.blob()]);
    }
    if (contentType && contentType.includes('application/json')) {
      return Promise.all([response.ok, response.json()]);
    }
    // it includes all text types
    if (contentType && contentType.includes('text/')) {
      return Promise.all([response.ok, response.text()]);
    }

    // unfortunately on download files there is no header available
    if (response.url && response.url.endsWith('.tgz') === true) {
      return Promise.all([response.ok, response.blob()]);
    }
  }

  return Promise.all([response.ok, response.text()]);
}

class API {
  public request<T>(url: string, method = 'GET', options: RequestInit = { headers: {} }): Promise<T> {
    if (!window.VERDACCIO_API_URL) {
      throw new Error('VERDACCIO_API_URL is not defined!');
    }

    const token = storage.getItem('token');
    const headers = new Headers(options.headers);

    if (token && headers.has('Authorization') === false) {
      headers.set('Authorization', `Bearer ${token}`);
      options.headers = headers;
    }

    if (!['http://', 'https://', '//'].some(prefix => url.startsWith(prefix))) {
      url = window.VERDACCIO_API_URL + url;
    }

    return new Promise((resolve, reject) => {
      fetch(url, {
        method,
        credentials: 'same-origin',
        ...options,
      })
        .then(handleResponseType)
        .then(response => {
          if (response[0]) {
            resolve(response[1]);
          } else {
            reject(new Error('something went wrong'));
          }
        })
        .catch(error => {
          reject(error);
        });
    });
  }
}

export default new API();
