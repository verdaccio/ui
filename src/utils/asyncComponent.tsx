import React from 'react';

export function asyncComponent(getComponent) {
  return class AsyncComponent extends React.Component {
    static Component = null;
    state = { Component: AsyncComponent.Component };

    componentDidMount() {
      const { Component } = this.state;
      if (!Component) {
        getComponent()
          .then(({ default: Component }) => {
            AsyncComponent.Component = Component;
            /* eslint react/no-did-mount-set-state:0 */
            this.setState({ Component });
          })
          .catch(err => {
            console.error(err);
          });
      }
    }
    render() {
      const { Component } = this.state;
      if (Component) {
        // eslint-disable-next-line verdaccio/jsx-spread
        // @ts-ignore
        return <Component {...this.props} />;
      }

      return null;
    }
  };
}
