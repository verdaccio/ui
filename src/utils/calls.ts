import { PackageMetaInterface } from '../../types/packageMeta';

import API from './api';

export async function callReadme(packageName: string, packageVersion?: string): Promise<string> {
  return await API.request<string>(
    `package/readme/${packageName}${packageVersion ? `?v=${packageVersion}` : ''}`,
    'GET'
  );
}

export async function callDetailPage(packageName: string, packageVersion?: string): Promise<PackageMetaInterface> {
  const packageMeta = await API.request<PackageMetaInterface>(
    `sidebar/${packageName}${packageVersion ? `?v=${packageVersion}` : ''}`,
    'GET'
  );

  return packageMeta;
}

export function callSearch(value: string, signal: AbortSignal) {
  // https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API#Browser_compatibility
  // FUTURE: signal is not well supported for IE and Samsung Browser
  return API.request(`search/${encodeURIComponent(value)}`, 'GET', { signal, headers: {} });
}
