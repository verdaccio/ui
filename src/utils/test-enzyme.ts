import { mount, shallow } from 'enzyme';
import React from 'react';

import ThemeProvider from 'verdaccio-ui/design-tokens/ThemeProvider';

const shallowWithTheme = (element: React.ReactElement<any>, ...props: any[]) =>
  shallow(element, {
    wrappingComponent: ThemeProvider,
    ...props,
  });

const mountWithTheme = (element: React.ReactElement<any>, ...props: any[]): any =>
  mount(element, {
    wrappingComponent: ThemeProvider,
    ...props,
  });

export { mountWithTheme as mount, shallowWithTheme as shallow };
