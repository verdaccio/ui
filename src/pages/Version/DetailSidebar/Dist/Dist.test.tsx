import React from 'react';

import { mount } from 'verdaccio-ui/utils/test-enzyme';

import { DetailContext } from '../../context';

import Dist from './Dist';

const withDistComponent = (packageMeta: React.ContextType<typeof DetailContext>['packageMeta']): JSX.Element => (
  <DetailContext.Provider value={{ packageMeta }}>
    <Dist />
  </DetailContext.Provider>
);

describe('<Dist /> component', () => {
  test('should render the component in default state', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: '',
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render the component with license as string', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: 'MIT',
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render the component with license as object', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: {
          type: 'MIT',
          url: 'https://www.opensource.org/licenses/mit-license.php',
        },
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });
});
