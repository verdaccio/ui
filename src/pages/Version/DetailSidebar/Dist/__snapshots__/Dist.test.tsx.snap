// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dist /> component should render the component in default state 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<body>
    <div>
      <ul
        class="MuiList-root MuiList-padding MuiList-subheader"
      >
        <span
          class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
        >
          Latest Distribution
        </span>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                file count
              </b>
              : 
              7
            </span>
          </div>
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                Size
              </b>
              : 
              10.00 Bytes
            </span>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </body>,
  "container": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<div>
    <ul
      class="MuiList-root MuiList-padding MuiList-subheader"
    >
      <span
        class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
      >
        Latest Distribution
      </span>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              file count
            </b>
            : 
            7
          </span>
        </div>
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              Size
            </b>
            : 
            10.00 Bytes
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Dist /> component should render the component with license as object 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<body>
    <div>
      <ul
        class="MuiList-root MuiList-padding MuiList-subheader"
      >
        <span
          class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
        >
          Latest Distribution
        </span>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                file count
              </b>
              : 
              7
            </span>
          </div>
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                Size
              </b>
              : 
              10.00 Bytes
            </span>
          </div>
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                License
              </b>
              : 
              MIT
            </span>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </body>,
  "container": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<div>
    <ul
      class="MuiList-root MuiList-padding MuiList-subheader"
    >
      <span
        class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
      >
        Latest Distribution
      </span>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              file count
            </b>
            : 
            7
          </span>
        </div>
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              Size
            </b>
            : 
            10.00 Bytes
          </span>
        </div>
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              License
            </b>
            : 
            MIT
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Dist /> component should render the component with license as string 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<body>
    <div>
      <ul
        class="MuiList-root MuiList-padding MuiList-subheader"
      >
        <span
          class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
        >
          Latest Distribution
        </span>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                file count
              </b>
              : 
              7
            </span>
          </div>
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                Size
              </b>
              : 
              10.00 Bytes
            </span>
          </div>
          <div
            class="MuiChip-root emotion-4 emotion-5"
          >
            <span
              class="MuiChip-label"
            >
              <b>
                License
              </b>
              : 
              MIT
            </span>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </body>,
  "container": .emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding-left: 0;
  padding-right: 0;
}

.emotion-4 {
  margin-right: 5px;
  text-transform: capitalize;
}

<div>
    <ul
      class="MuiList-root MuiList-padding MuiList-subheader"
    >
      <span
        class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
      >
        Latest Distribution
      </span>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              file count
            </b>
            : 
            7
          </span>
        </div>
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              Size
            </b>
            : 
            10.00 Bytes
          </span>
        </div>
        <div
          class="MuiChip-root emotion-4 emotion-5"
        >
          <span
            class="MuiChip-label"
          >
            <b>
              License
            </b>
            : 
            MIT
          </span>
        </div>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
